mantidproject/mantid
# SANSILL process solvent

This PR adds `solvent` as another correction that can be subtracted from the corrected 2D sample data. The solvent is processed separately as a sample, in the same way as the water reference. Then, it is subtracted from the sample data after all other corrections are imposed and after the water reference. 

The changes include a new property in `SANSILLReduction`, `SolventInputWorkspace` that if specified, will be deducted from data. `SANSILLAutoProcess` also has a new property, `SolventFiles`, that is used to provide the processed solvent workspaces to `SANSILLReduction`. An additional column for the `SolventFiles` has been also added to `DrILL` interface rundex.

**To test:**


1. Make sure D11 SystemTest data is in your user data search directories.
2. From the workbench, run  the following script:

```
from mantid.simpleapi import *
from tempfile import gettempdir

MaskBTP(Instrument='D11', Tube='0-6,250-256')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_vertical')
MaskBTP(Instrument='D11', Pixel='0-4,252-256')
Plus(LHSWorkspace='mask_vertical', RHSWorkspace='D11MaskBTP', OutputWorkspace='edge_masks')
# the edges mask can be used as a default mask for all distances and wavelengths

MaskBTP(Instrument='D11', Tube='115-138,', Pixel='115-139')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_8m_4_6A_center')
MaskBTP(Instrument='D11', Tube='0-14', Pixel='241-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_8m_4_6A_center', OutputWorkspace='mask_8m_4_6A')

MaskBTP(Instrument='D11', Tube='110-146', Pixel='110-146')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_1m_4_6A_center')
MaskBTP(Instrument='D11', Tube='3-14', Pixel='240-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_1m_4_6A_center', OutputWorkspace='mask_1m_4_6A_int')
MaskBTP(Instrument='D11', Tube='77-107', Pixel='245-256')
Plus(LHSWorkspace='D11MaskBTP', RHSWorkspace='mask_1m_4_6A_int', OutputWorkspace='mask_1m_4_6A')

beams = '947,1088'
containers = '973,1003'
container_tr = '988'
beam_tr = '1119'
samples = '975,1005'
sample_tr = '990'
solvents = '1106,1091'
solvent_tr = '1121'
thick = 0.1

# this also tests that already loaded workspace can be passed instead of a file
LoadNexusProcessed(Filename='sens-lamp.nxs', OutputWorkspace='sens-lamp')
# first, process the solvent
SANSILLAutoProcess(
            SampleRuns=solvents,
            BeamRuns=beams,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=solvent_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='solvents',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05,
            ClearCorrected2DWorkspace=False,
            StitchReferenceIndex=0
)

tmp_dir = gettempdir()
solvent_dir = [os.path.join(tmp_dir, 'solvent_' + str(i) + '.nxs') for i in range(2)]
SaveNexusProcessed('001106_Sample', solvent_dir[0])
SaveNexusProcessed('001091_Sample', solvent_dir[1])

# reduce samples
SANSILLAutoProcess(
            SampleRuns=samples,
            BeamRuns=beams,
            ContainerRuns=containers,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=sample_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SolventFiles=",".join(solvent_dir),
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='iq_mult_solvent',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05,
            StitchReferenceIndex=0
)
```
3. Make sure it runs correctly, that the solvent I(Q) and corrected 2D solvent data is created and that it is subtracted correctly from the sample data.


Fixes #30612 
Part of #30343  

Release notes added.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
